### PR TITLE
CI: run mypy in full env

### LINKED ISSUE #4881: check mypy at the end of some CI runs?
We currently run mypy in the `pre-commit hooks` CI. However, this is done in an environment where no dependencies are installed. Therefore we missed the errors that pop up when running mypy with numpy 1.20 installed. (Please correct my if I misunderstood this). Should we add a new step to our CI and run mypy?

I think we should at least add this to `ubuntu-latest py3.8`. For more complete checks we could also go for ` ubuntu-latest py37-min-all-deps` and `upstream-dev`.

### PR DESCRIPTION
**Pull request by mathause**
- [x] Closes #4881
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

I only added one run for py3.8 latest. To be entirely sure we could also check the typing `py37-min-all-deps` but it feels like overkill...

---

Ok, looks good - the failure is expected - see #4878.

### CODE REVIEW

**Comment by keewis** [comment]
the workflow looks good to me, but it should probably be possible to skip it using `[skip-ci]`

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -30,6 +30,7 @@ repos:
   #     - id: velin
   #       args: ["--write", "--compact"]
   - repo: https://github.com/pre-commit/mirrors-mypy
+    # version must correstpond to the one in .github/workflows/ci-additional.yaml
```
Comment (keewis) [inline .pre-commit-config.yaml:33]:
```suggestion
    # version must correspond to the one in .github/workflows/ci-additional.yaml
```

File: .github/workflows/ci-additional.yaml [hunk 2]
```diff
@@ -185,7 +187,7 @@ jobs:
         run: |
           mamba env update -f ci/requirements/environment.yml
       - name: Install mypy
-        run: |
+        run: | # version must correstpond to the one in .pre-commit-config.yaml
           mamba install mypy=0.800
```
Comment (keewis) [inline .github/workflows/ci-additional.yaml]:
```suggestion
        run: |
          # version must correspond to the one in .pre-commit-config.yaml
          mamba install mypy=0.800
```

File: .github/workflows/ci-additional.yaml [hunk 3]
```diff
@@ -157,6 +157,50 @@ jobs:
...
+          mamba-version: "*"
+          activate-environment: xarray-tests
+          auto-update-conda: false
+          python-version: "3.8"
+
+      - name: Install conda dependencies
+        run: |
+          mamba env update -f ci/requirements/environment.yml
+      - name: Install mypy
+        run: | # version must correstpond to the one in .pre-commit-config.yaml
```
Comment (mathause) [inline .github/workflows/ci-additional.yaml:190]:
```suggestion
        run: |
        # version must correspond to the one in .pre-commit-config.yaml
```

File: setup.cfg [hunk 4]
```diff
@@ -165,6 +165,7 @@ default_section = THIRDPARTY
 known_first_party = xarray
 
 [mypy]
+show_error_codes = True
```
Comment (mathause) [inline setup.cfg:168]:
```suggestion
```

File: .github/workflows/ci-additional.yaml [hunk 5]
```diff
@@ -157,6 +157,51 @@ jobs:
...
+          activate-environment: xarray-tests
+          auto-update-conda: false
+          python-version: "3.8"
+
+      - name: Install conda dependencies
+        run: |
+          mamba env update -f ci/requirements/environment.yml
+      - name: Install mypy
+        run: |
+        # version must correspond to the one in .pre-commit-config.yaml
```
Comment (mathause) [inline .github/workflows/ci-additional.yaml:191]:
```suggestion
          # version must correspond to the one in .pre-commit-config.yaml
```

File: .github/workflows/ci-additional.yaml [hunk 6]
```diff
@@ -157,6 +157,51 @@ jobs:
         run: |
           python -m pytest --doctest-modules xarray --ignore xarray/tests
 
+  typing:
+    name: Type checking (mypy)
+    runs-on: "ubuntu-latest"
+    needs: detect-ci-trigger
+    if: needs.detect-ci-trigger.outputs.triggered == 'false'
```
Comment (keewis) [inline .github/workflows/ci-additional.yaml:164]:
```suggestion
    if: false && needs.detect-ci-trigger.outputs.triggered == 'false'
```

**Review by keewis — APPROVED** [review]
looks good to me. Before merging, I think we should disable the CI for now, though, please enable it in #4878

**Review by andersy005 — APPROVED** [review]

File: .github/workflows/ci-additional.yaml [hunk 7]
```diff
@@ -157,6 +157,51 @@ jobs:
...
+          auto-update-conda: false
+          python-version: "3.8"
+
+      - name: Install conda dependencies
+        run: |
+          mamba env update -f ci/requirements/environment.yml
+      - name: Install mypy
+        run: |
+          # version must correspond to the one in .pre-commit-config.yaml
+          mamba install mypy=0.800
```
Comment (andersy005) [inline .github/workflows/ci-additional.yaml:192]: @keewis, is there any way we can automatically keep this version in sync with changes that will be introduced by https://github.com/pydata/xarray/blob/master/.github/workflows/ci-pre-commit-autoupdate.yaml? In other words, can we make it easy to update this version whenever `.pre-commit-config.yaml` is updated?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mathause) [inline .pre-commit-config.yaml:33]:
```suggestion
    # version must correspond to the one in .github/workflows/ci-additional.yaml
```

File: .github/workflows/ci-additional.yaml (hunk 7 above)
Comment (keewis) [reply .github/workflows/ci-additional.yaml]: sure. The action used to create the PR allows specifying any number of commands, so we can add a script which does the syncing. That might be easier if we create a new environment file (`requirements.txt` like?) containing just `mypy`, though.

**Comment by keewis** [comment]
Thanks, @mathause, looks good to me.

I wonder if we should move this into `linting`? That way, running the linting on autoupdate PRs would be easier. That should probably be a new PR, though, so I'm merging.

**Comment by mathause** [comment]
Thanks - just for reference - this is currently unconditionally inactive & should be activated in #4878

**Pull request merged**